pennylane/device_tracker.py

```diff
@@ -0,0 +1,259 @@
+        self._time_last = time.time()
+
+class PrintTotals(DefaultTracker):
+
+    def record(self):
+        """Print all key-value pairs stored in ``totals``.
+        """
+        print("Total: ", end="")
+        for key, value in self.totals.items():
+            print(f"{key} = {value}", end="\t")
+        print()
+
+class PrintCurrent(DefaultTracker):
+
+    def record(self):
+        """Print all key-value pairs stored in ``current``.
+        """
+        print("Current: ", end="")
+        for key, value in self.current.items():
+            print(f"{key} = {value}", end="\t")
+        print()
+
+class PrintCustom(DefaultTracker):
+    def __init__(self, dev=None, reset_on_enter=True, custom_recorder=None):
+        self.custom_recorder=custom_recorder
+        super(PrintCustom, self).__init__(dev=dev, reset_on_enter=reset_on_enter)
+
+    def record(self):
+        """Executes user-provided record function."""
+        self.custom_recorder(totals=self.totals, history=self.history, current=self.current,)
+
+record_mapping = {"totals": PrintTotals, "current": PrintCurrent, "custom": PrintCustom}
+update_mapping = {"timings": UpdateTimings}
+
+def track(dev, record=None, update=None, **kwargs):
+    r"""Creates a tracking context and applies it to a device.
+
+    Args:
+        dev (Device): a PennyLane-compatible device
+        record (callable or str or None): If callable, this function is used to record information. Must be a
+            function of ``current``, ``totals`` and ``history`` keywords. If string, selects an built record method.
+            Current available options are ``"totals"`` and ``"current"``. If ``None``, no recording happens.
+        update (str or None): if ``"timings"``, the update method will also store the length of system time between
+            subsequent `update` calls.
+
+    Keyword Args:
+        reset_on_enter=True (bool): whether or not to reset information
+            entering the context
+
+    **Example**
+
+    With the default settings on most devices, the tracker will store execution and shot information without 
+    printing or logging of the information.  This information can be accessed through the `totals`, `history`,
+    and `current` attributes of the tracker.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1, shots=100)
+        
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+        
+        x = np.array(0.1)
+
+        with qml.track(dev) as tracker:
+            qml.grad(circuit)(0.1)
+
+        You can then access the tabulated information through ``totals``, ``history``, and ``current``:
+
+        >>> tracker.totals
+        {'executions': 3, 'shots': 300}
+        >>> tracker.history
+        {'executions': [1, 1, 1], 'shots': [100, 100, 100]}
+        >>> tracker.current
+        {'executions': 1, 'shots': 100}
```

Comment:
If someone wants to log the last thing that was passed to update.  We can't use history, because it may have keywords not present during the last update.

Reply:
Renamed to `tracker.latest`.  Hoping that word is better.